File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/NullCheckExpression.java
package me.itzisonn_.meazy.parser.ast.expression;

import lombok.Getter;
import me.itzisonn_.meazy.instruction.InstructionsSet;
import me.itzisonn_.meazy.parser.ast.ProgramUnit;
import me.itzisonn_.meazy.parser.DataType;
import me.itzisonn_.meazy.runtime.environment.Environment;
import me.itzisonn_.meazy.util.MiscUtils;
import org.jspecify.annotations.NullMarked;

import java.lang.constant.ClassDesc;
import java.util.UUID;

@Getter
@NullMarked
public class NullCheckExpression implements Expression {
    private final Expression checkExpression;
    private final Expression nullExpression;

    public NullCheckExpression(Expression checkExpression, Expression nullExpression) {
        this.checkExpression = checkExpression;
        this.nullExpression = nullExpression;
    }

    @Override
    public void emit(InstructionsSet instructionsSet, Environment environment, ProgramUnit parent) {
        DataType checkExpressionType = checkExpression.getType(environment, this);
        if (!checkExpressionType.isNullable()) {
            checkExpression.emit(instructionsSet, environment, this);
            return;
        }

        var endLabel = instructionsSet.createAndInitLabel();

        checkExpression.emit(instructionsSet, environment, this);
        instructionsSet.duplicate();
        instructionsSet.gotoLabelIfNonNull(endLabel);

        instructionsSet.pop();
        nullExpression.emit(instructionsSet, environment, this);
        ClassDesc nullExpressionClassDesc = nullExpression.getType(environment, this).getClassDesc();
        if (nullExpressionClassDesc.isPrimitive()) MiscUtils.boxPrimitive(instructionsSet, nullExpressionClassDesc);
        instructionsSet.gotoLabel(endLabel);

        instructionsSet.bindLabel(endLabel);
    }

    @Override
    public DataType getType(Environment environment, ProgramUnit parent) {
        DataType checkExpressionType = checkExpression.getType(environment, this);
        if (!checkExpressionType.isNullable()) return checkExpressionType;

        DataType nullExpressionType = nullExpression.getType(environment, this);
        ClassDesc nullExpressionClassDesc = nullExpressionType.getClassDesc();
        if (nullExpressionClassDesc.isPrimitive()) nullExpressionClassDesc = MiscUtils.getBoxedType(nullExpressionClassDesc);

        return DataType.commonOf(environment, checkExpressionType.asNonNull(), nullExpressionType.with(nullExpressionClassDesc));
    }
}

File: src/main/java/me/itzisonn_/meazy/parser/ast/expression/NullCheckExpression.kt
package me.itzisonn_.meazy.parser.ast.expression

import me.itzisonn_.meazy.instruction.InstructionsSet
import me.itzisonn_.meazy.parser.DataType
import me.itzisonn_.meazy.parser.ast.ProgramUnit
import me.itzisonn_.meazy.runtime.environment.Environment
import me.itzisonn_.meazy.util.MiscUtils.boxPrimitive
import me.itzisonn_.meazy.util.MiscUtils.getBoxedType

class NullCheckExpression(
    val checkExpression: Expression,
    val nullExpression: Expression
) : Expression {
    override fun emit(instructions: InstructionsSet, environment: Environment, parent: ProgramUnit) {
        val checkExpressionType = checkExpression.getType(environment, this)
        if (!checkExpressionType.isNullable) {
            checkExpression.emit(instructions, environment, this)
            return
        }

        val endLabel = instructions.createAndInitLabel()

        checkExpression.emit(instructions, environment, this)
        instructions.duplicate()
        instructions.gotoLabelIfNonNull(endLabel)

        instructions.pop()
        nullExpression.emit(instructions, environment, this)
        val nullExpressionClassDesc = nullExpression.getType(environment, this).classDesc
        if (nullExpressionClassDesc.isPrimitive) boxPrimitive(instructions, nullExpressionClassDesc)
        instructions.gotoLabel(endLabel)

        instructions.bindLabel(endLabel)
    }

    override fun getType(environment: Environment, parent: ProgramUnit): DataType {
        val checkExpressionType = checkExpression.getType(environment, this)
        if (!checkExpressionType.isNullable) return checkExpressionType

        val nullExpressionType = nullExpression.getType(environment, this)
        var nullExpressionClassDesc = nullExpressionType.classDesc
        if (nullExpressionClassDesc.isPrimitive) nullExpressionClassDesc = getBoxedType(nullExpressionClassDesc)

        return DataType.commonOf(
            environment,
            checkExpressionType.asNonNull(),
            nullExpressionType.with(nullExpressionClassDesc)
        )
    }
}
